### DISCUSSION TITLE
Increase the precision used for `string_of_float`

### DISCUSSION THREAD

**Discussion by dra27** [ocaml/ocaml#11975] [Features]
Apropos https://github.com/ocaml/ocaml/pull/11353#issuecomment-1165585708, we should experiment with `%.17g` for `string_of_float`.

**Comment by craff** [comment]
Hello,

I read in postgresql documentation that 17 digit allows to give a decimal representation that will be guaranty to be rounded 
exactly to the original double. This would ensure that string_of_float / float_of_string are really inverse function and I think this should be the default.  Would %.17g ensure this ?

postgresql is even avoiding printing a decimal value just in between two floats because some platform get the rounding wrong apparently.

I quote postgresql [documentation](https://www.postgresql.org/docs/current/datatype-numeric.html)

> By default, floating point values are output in text form in their shortest precise decimal representation; the decimal value produced is closer to the true stored binary value than to any other value representable in the same binary precision. (However, the output value is currently never exactly midway between two representable values, in order to avoid a widespread bug where input routines do not properly respect the round-to-nearest-even rule.) This value will use at most 17 significant decimal digits for float8 values, and at most 9 digits for float4 values.
> Note
> 
> This shortest-precise output format is much faster to generate than the historical rounded format.

**Comment by craff** [comment]
The following code gives some hint that .17g seems to behave as postgresql doc suggest:
After a rather long run, it still did not find any counter-example.

```
let to_string x = Printf.sprintf "%.17g" x
let of_string x = float_of_string x

let test () =
  while true do
    let x = Random.float 1.0 in
    if of_string (to_string x) <> x then
      Printf.printf "bad float: %.13h\n" x
  done

let _ = test ()
```

**Comment by craff** [comment]
The converse `to_string (of_string x)` is certainly wrong, but it is not the important property that we want to ensure.
Remark: postgres does not support hexadecimal litteral double, and only support binary echange to get the result, so as far as data base interface is to consider, it is important to have a safe conversion between float and string and string_of_float should
be safe by default.

**Comment by xavierleroy** [comment]
Yes, there's general agreement that `%.17g` produces a representation that always reads back to the original FP number.  However, all the C standard libraries that I tested have a suboptimal implementation of `%.17g` that produces much more digits than required.  For example, with Glibc,
```
1 / 10 = 0.10000000000000001
2 / 10 = 0.20000000000000001
3 / 10 = 0.29999999999999999
4 / 10 = 0.40000000000000002
5 / 10 = 0.5
6 / 10 = 0.59999999999999998
7 / 10 = 0.69999999999999996
8 / 10 = 0.80000000000000004
9 / 10 = 0.90000000000000002
```
while Java's `toString` method (which is required by the Java standard to produce minimal outputs) correctly gives
```
1 / 10 = 0.1
2 / 10 = 0.2
3 / 10 = 0.3
4 / 10 = 0.4
5 / 10 = 0.5
6 / 10 = 0.6
7 / 10 = 0.7
8 / 10 = 0.8
9 / 10 = 0.9
```
Printing 0.59999999999999998 when 0.6 would do is amateurish.  Even 1980's pocket calculators did better than that.

**Comment by craff** [comment]
OCaml is ok there, I checked this no problem both for unwanted representation and reading back. 
But we need a theorem not a "general agreement" ?

```

let to_string x = Printf.sprintf "%.17g" x
let of_string x = float_of_string x

let check x =
  if of_string (to_string x) <> x then
    Printf.printf "bad float: %.13h\n" x

let random_test () =
  while true do
    check (Random.float 1.0)
  done

let _ =
  for j = 1 to 20 do
    for i = 0 to 10 do
      let x = float i /. (10. ** float j) in
      Printf.printf "%17g\n" x;
      check x;
    done
  done
```

**Comment by craff** [comment]
For a theorem, if we can not find it, we must ask [Jean-Michel Muller](https://perso.ens-lyon.fr/jean-michel.muller/), I am pretty sure he will have something to say ?

**Comment by craff** [comment]
It's done, I sent him an email ...

**Comment by dra27** [comment]
> OCaml is ok there

What do you mean by OK? I'm seeing the same as @xavierleroy on Linux, even before daring to see what ancient MSVCRTs do with this:

```
# string_of_float 0.1;;
- : string = "0.1"
# Printf.printf "%.17g\n%h\n%h\n" 0.1 0.1 0.10000000000000001;;
0.10000000000000001
0x1.999999999999ap-4
0x1.999999999999ap-4
```

**Comment by dra27** [comment]
Prior art: [postgres/postgres#02ddd499](https://github.com/postgres/postgres/commit/02ddd499322ab6f2f0d58692955dc9633c2150fc) and the code in [ulfjack/ryu](https://github.com/ulfjack/ryu/) on which it's based. The JDK implementation looks similarly enormous!

**Comment by xavierleroy** [comment]
@craff: you are still fixated on the correctness condition (the round-trip float->string->float property), but now we're discussing optimality (minimal number of digits).  There's an extensive literature about correct and optimal algorithms for FP -> string conversion, and a number of implementations of these algorithms.  The problem is that these implementations are quite large, don't produce their output in %g format (usually), and are not provided by the C standard libraries like they should.

**Comment by craff** [comment]
Sorry, my test code above is buggy: I wrote %17g instead of %.17g in the latest loop. 

I agree that 0.10000000000000001 is enoying.

**Comment by craff** [comment]
Is this kind of idea below completely stupid ?

If is (surprisingly) only twice as slow as ".17g", it can be read back, does the best for decimal numbers.
I guess that written in C and may be optimized for the precision of the initial value of m it could be even faster than
".17g".

```
let string_of_float x =
  let neg = x < 0.0 in
  let x = abs_float x in
  let e = ref (int_of_float (Float.ceil (log10 x)) - 17) in
  (* NOTE: maybe use libc exp10l would be more precise? libc doc
     says "mathematically the same" ... but does it means the same ? *)
  let m = ref (int_of_float (Float.round (x *. (10. ** (float (- !e)))))) in
  if !m >= 100_000_000_000_000_000 then (m := !m / 10; incr e);
  if !m <   10_000_000_000_000_000 then (m := !m * 10; decr e);
  let m100 = !m mod 100 in
  let ma = if m100 > 50 then !m + 100 - m100 else !m - m100 in
  let ra = Printf.sprintf "%de%d" ma !e in
  let ya = float_of_string ra in
  if x = ya then
    begin
      m := ma;
      while !m mod 10 = 0 && !m > 0 do
        m := !m / 10; incr e
      done;
      Printf.sprintf "%s%de%d" (if neg then "-" else "") !m !e
    end
  else begin
    let r = ref (Printf.sprintf "%de%d" !m !e) in
    let y = ref (float_of_string !r) in
    while x <> !y do
      if !y < x then incr m else decr m;
      r := Printf.sprintf "%de%d" !m !e;
      y := float_of_string !r;
      (*Printf.printf "%dE%d %.13h %.13h\n%!" !m !e !y x;*)
    done;
    if neg then "-" ^ !r else !r;
    end
```

**Comment by craff** [comment]
By the way it calls in average 2.14 times float_of_string per call for random floats.

**Comment by craff** [comment]
Jean Michel Muller pointed me to Gay's code here:
https://github.com/jwiegley/gdtoa

It is big, but well tested and robust. We may particularise the code for mode 0 "shortest output". 
which should simplify it a bit (at the risk of breaking it) it gives an algorithm with the shortest representation
that allows the roundtrip. May be a good default for string_of_float.

I tested in C:
printf(".17g",...) => 0.274151s
g_dfmt from Gay's code with ndig=0 (shortest exact) => 0.531940s (which produces %g format)

So two times slower but very good properties ....
particularizing the code to ndig=0 should help a bit.

**Comment by dbuenzli** [comment]
> Jean Michel Muller pointed me to Gay's code here:

#7218 also has a few pointers to other alternate implementations.

**Comment by xavierleroy** [comment]
See Ryu, the best(?) known algorithm for float-string conversion: https://dl.acm.org/doi/10.1145/3296979.3192369

Here is an implementation: https://github.com/ulfjack/ryu .  However, the result is in `%e` format (e.g. `1.1e-1`), and a little more work is needed to get `%g` format (e.g. `0.11`).

**Comment by gasche** [comment]
> A little more work is needed to get %g format (e.g. 0.11)

According to the Ryu readme:

> In most implementations, printf provides three floating-point specific formatters, %f, %e, and %g:
>
>- The %f format prints the full decimal part of the given floating point number, and then appends as many digits of the fractional part as specified using the precision parameter.
>
> - The %e format prints the decimal number in scientific notation with as many digits after the initial digit as specified using the precision parameter.
>
> - The %g format prints either %f or %e format, whichever is shorter.
>
> Ryu Printf implements %f and %e formatting in a way that should be drop-in compatible with most implementations of printf, although it currently does not implement any formatting flags other than precision

(On my system `man 3 printf` gives a specification that is more complex than "whichever of %f and %e is shorter".)

Given that the speedups (compared to tested libc implementations) are at least 4x, it would probably be fine performance-wise to always ask Ryu to produce both %f and %e, and then return the shortest output.

(I think the question is rather the intimidating complexity of those implementations.)

**Comment by craff** [comment]
On 22-07-20 02:48:28, Xavier Leroy wrote:
> See Ryu, the best(?) known algorithm for float-string conversion: https://
> dl.acm.org/doi/10.1145/3296979.3192369
>
> Here is an implementation: https://github.com/ulfjack/ryu . However, the result
> is in %e format (e.g. 1.1e-1), and a little more work is needed to get %g
> format (e.g. 0.11).
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.*Message ID: <ocaml/ocaml/
> ***@***.***>
>

I may write a as clean as possible implementation supporting e,f and g format.
allowing something like "%.$[fge]" to specify shortest representation that reads backs
exactly in all printf formats. As I said before, this is needed for data base interface
which do not support hexadecimal float representation.

This are some questions about it:

1) Do we accept a table of approximately 600 (64 bits mantissa + 16 bits
exponent) to precompute all power of 10 with some extra precision ?
It seem to me that it allows for a simpler/faster algorithm, but will slow
down the start of any ocaml program and take a few Kb ?

Ryu's algorithm (or another one, I may mix up things) use a partial table,
why not a full one ?

2) ensuring readback means we know that the current float_of_string take the
nearest float on all platforms. Can we assume this or should we also change
this implementation ? I am worry about the remark in pgsql doc where they
volontarely avoid case of tie because they say some implementations get the
rounding in the wrong direction ?


Cheers,
Christophe

--
Christophe Raffalli
tél: +689 87 23 11 48
web: http://raffalli.eu

**Comment by zimmermann6** [comment]
first it is indeed a theorem that printing a double-precision number with 17 significant digits will enable to read it back exactly (assuming both conversions are correctly rounded, with rounding to nearest). See for example page 117 of "Modern Computer Arithmetic" (https://members.loria.fr/PZimmermann/mca/mca-cup-0.5.9.pdf).

About the smallest output that when read back produces the original double precision value, there is an efficient algorithm "Schubfach" by Raffaelo Giulietti (https://drive.google.com/file/d/1luHhyQF9zKlM8yJ1nebU0OgVYhfC6CBN/view) that we reviewed with Jean-Michel Muller and that is now implemented in OpenJDK (https://github.com/openjdk/jdk/pull/3402). I attach our review.

[review.pdf](https://github.com/ocaml/ocaml/files/10336455/review.pdf)

**Comment by xavierleroy** [comment]
Thanks for the pointer to the Schubfach algorithm.  Do you know if there's an open-source C implementation available somewhere? preferably not too big?  Some of the reasons why we've not adopted Ryu or other clever algorithms yet is that available implementations tend to be large and/or rely on C++ or Java library functions.

**Comment by zimmermann6** [comment]
I don't know of any open-source C implementation, but you can ask Raffaelo Giulietti

**Comment by mroch** [comment]
if it helps, we have an MIT-licensed C port of google's double-conversion (grisu) here: https://github.com/flow/ocaml-dtoa

i don't know whether you'd call it large.

one small thing I like about it is that all the JS engines use double-conversion, so the js_of_ocaml implementation is to just call the native JS `.toString` 🤷

**Comment by xavierleroy** [comment]
Should this and related issues be moved to a discussion?  We have to make up our minds between 1- inaccurate printing (`%.12g`), 2- printing with too many digits (`%.17g`), and 3- integrating big chunks of third-party C code in the OCaml runtime system to get minimal exact output.  I don't see how we're going to choose.

**Comment by craff** [comment]
Hello, There is a short code that I tested that find an exact solution by trial and error. It is necessarily slow. I will try to find back this code...

Le 29 janvier 2023 08:32:54 GMT-10:00, Xavier Leroy ***@***.***> a écrit :

**Comment by zimmermann6** [comment]
Hi,

the only definitive answer is to get correctly rounded output (minimal or not).
Thus 2 or 3 would be fine to me (assuming 2 is correctly rounded too).

Paul
